[Serve] Improve `_best_fit_node` logic to consider prioritized custom resources

ktyxx opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR enhances the logic of DeploymentScheduler._best_fit_node() to consider custom resource prioritization defined via the RAY_SERVE_CUSTOM_RESOURCES environment variable. The updated logic ensures that nodes are selected not just based on generic resources (CPU, GPU, memory), but also takes into account the specified order of custom resources when minimizing resource fragmentation.

This change addresses inefficient scheduling behavior where custom resources like GRAM were ignored in prioritization, resulting in unnecessary fragmentation (e.g., #51361).
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
Closes #51361
<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ktyxx committed:
[serve] Update _best_fit_node in DeploymentScheduler to choose the best-fit node based on custom resources

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx committed:
Format code with script

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx committed:
Fix: add missing os import to satisfy ruff lint

Signed-off-by: kitae <ryugitae777@gmail.com>

abrarsheikh commented on python/ray/serve/_private/deployment_scheduler.py:
```diff
@@ -38,6 +39,11 @@ class SpreadDeploymentSchedulingPolicy:
 
 @total_ordering
 class Resources(dict):
+    # Custom resource priority from environment variable
+    CUSTOM_PRIORITY = [
+        r for r in os.environ.get("RAY_SERVE_CUSTOM_RESOURCES", "").split(",") if r
+    ]
+
```
let's add this to `python/ray/serve/_private/constants.py,` that way; all constants are in one place.

can we find a different name for env var, `RAY_SERVE_CUSTOM_RESOURCES` does not convey the intent here. Maybe `RAY_SERVE_HIGH_PRIORITY_RESOURCES`?

abrarsheikh replied:
good point, another suggestion, `RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES` signals that a subset of custom resources will have the highest priority.

ktyxx replied:
Changes applied as discussed — would appreciate another review when you have time

abrarsheikh commented on python/ray/serve/tests/unit/test_deployment_scheduler.py:
```diff
@@ -433,6 +434,30 @@ def test_best_fit_node():
     )
 
+    # Custom resource prioritization: customx is more important than customy
+    os.environ["RAY_SERVE_CUSTOM_RESOURCES"] = "customx,customy"
```
let's use the mock. patch or something similar here to patch the value of the env variable; this will ensure the environment variable is properly reset across tests.

ktyxx replied:
I already included a manual restore step with Resources.CUSTOM_PRIORITY = original, but I agree that using mock.patch.dict to temporarily override the environment variable would be cleaner and safer, especially in case of test failures or future test additions.

I'll update the test to use that approach instead.

abrarsheikh commented:
Minor tangent: What do you think about changing the implementation to allow full control over the priority order of all resources via a single environment variable? For example:

```bash
RAY_SERVE_RESOURCE_PRIORITY='r1,CPU,GPU,r2'
```

The default value could be 'GPU,CPU,memory' to maintain backward compatibility. This approach keeps the implementation generic while still meeting your specified requirements.

ktyxx committed:
refactor(scheduler): rename custom resource priority env var and apply mock in test

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx commented:
@abrarsheikh 
Your suggestion makes a lot of sense and I agree with the rationale behind it.

That said, my intention with introducing custom_resource prioritization was to address cases where users explicitly care about their custom resource types.

Allowing full control over general resources like GPU, CPU, and memory might introduce additional complexity, and in many cases, similar prioritization behavior can still be achieved by adjusting the relative values of custom resources.

So I'd prefer to keep the scope focused on custom resources for now, but I’m open to revisiting the broader generalization if we see strong use cases for it.

abrarsheikh commented on python/ray/serve/_private/deployment_scheduler.py:
```diff
@@ -38,6 +40,11 @@ class SpreadDeploymentSchedulingPolicy:
 @total_ordering
 class Resources(dict):
+    # Custom resource priority from environment variable
+    CUSTOM_PRIORITY = [
+        r for r in RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES.split(",") if r
+    ]
```
nit: I would prefer the `.split` logic to be `constants.py` purely for consistency with other constants in the file.

zcin replied:
+1

abrarsheikh commented on python/ray/serve/tests/unit/test_deployment_scheduler.py:
```diff
@@ -433,6 +437,33 @@ def test_best_fit_node():
     )
 
+    # Custom resource prioritization: customx is more important than customy
+    with mock.patch.dict(os.environ, {RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES: "customx,customy"}):
+        original = Resources.CUSTOM_PRIORITY
+        Resources.CUSTOM_PRIORITY = [
```
the reason why the patch on `os.environ` is not taking effect is that by the time this test is run the class is already defined and `Resources.CUSTOM_PRIORITY` gets set. I suggest the following instead

1. direct use `RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES` inside `__lt__` function causing its value to be resolved at runtime
2. patch the value of `RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES`


```
with mock.patch('ray.serve._private.deployment_scheduler.RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES', ['customx', customy]):
	# test

```

ktyxx replied:
Thank you for the suggestions! I’ve applied the changes as recommended. Appreciate your feedback

abrarsheikh commented:
@zcin can you take a look

zcin commented on python/ray/serve/_private/constants.py:
```diff
@@ -376,6 +376,11 @@ def parse_latency_buckets(bucket_str: str, default_buckets: list) -> list:
     os.environ.get("RAY_SERVE_USE_COMPACT_SCHEDULING_STRATEGY", "0") == "1"
 )
 
+# Comma-separated list of custom resources prioritized in scheduling (e.g., "customx,customy")
```
```suggestion
# Comma-separated list of custom resources prioritized in scheduling. Sorted from highest to lowest priority.
# Example: "customx,customy"
```

zcin commented on python/ray/serve/_private/deployment_scheduler.py:
```diff
@@ -77,18 +84,23 @@ def __sub__(self, other):
-        sorted higher than a resource with a smaller number of GPUs,
-        regardless of the values of the other resource types. Similarly
-        for CPU next, memory next, etc.
+        1. Custom resources defined in RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES (in order)
```
```suggestion
        1. Custom resources defined in RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES (sorted by priority)
```

zcin commented on python/ray/serve/_private/deployment_scheduler.py:
```diff
@@ -38,6 +40,11 @@ class SpreadDeploymentSchedulingPolicy:
 @total_ordering
 class Resources(dict):
+    # Custom resource priority from environment variable
+    CUSTOM_PRIORITY = [
```
Can you add type hints

ktyxx replied:
Thank you for the suggestions! I’ve applied the changes as recommended. Appreciate your feedback

zcin reviewed (commented):
Overall LGTM, thanks for the contribution!

ktyxx committed:
[serve] Refactor: Move custom resource priority parsing into a function

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx committed:
[serve] Fix: Update comments for custom resource priority ordering

Signed-off-by: kitae <ryugitae777@gmail.com>

abrarsheikh commented on python/ray/serve/_private/constants.py:
```diff
@@ -376,6 +377,15 @@ def parse_latency_buckets(bucket_str: str, default_buckets: list) -> list:
     os.environ.get("RAY_SERVE_USE_COMPACT_SCHEDULING_STRATEGY", "0") == "1"
 )
 
+def parse_custom_resource_priority() -> List[str]:
+    return [r for r in RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES.split(",") if r]
+
+# Comma-separated list of custom resources prioritized in scheduling. Sorted from highest to lowest priority.
+# Example: "customx,customy"
+RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES = os.environ.get(
+    "RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES", ""
+)
```
```suggestion
def str_to_list(s: str) -> List[str]:
    l = [trim(r) for r in s.split(",") if r]
    return [e for e in l if e]

# Comma-separated list of custom resources prioritized in scheduling. Sorted from highest to lowest priority.
# Example: "customx,customy"
RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES = str_to_list(os.environ.get(
    "RAY_SERVE_HIGH_PRIORITY_CUSTOM_RESOURCES", ""
))
```

abrarsheikh replied:
I would also check if a str_to_list-like function exists in that constants file or if it already references a utility of that sorts

ktyxx committed:
refactor(scheduler): extract str_to_list for env parsing and clean up resource priority logic

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx commented:
Summary of changes
- Extracted str_to_list() to utils.py for consistent env var parsing
- Removed parse_custom_resource_priority() and updated usage accordingly
- Passed all pre-commit checks (ruff, black, etc.)
- Signed all commits with DCO

Ready for re-review.
@abrarsheikh Please take another look when you have time. Let me know if anything else is needed.

ktyxx committed:
Apply black formatting to constants.py

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx committed:
Merge remote-tracking branch 'upstream/master' into enhance-best-fit-node-logic

Signed-off-by: kitae <ryugitae777@gmail.com>

ktyxx committed:
Fix ruff formatting after merge

Signed-off-by: kitae <ryugitae777@gmail.com>

zcin reviewed (approved):
LGTM

abrarsheikh reviewed (approved):
nicely done. thanks for the contribution.

zcin committed:
Merge branch 'master' into enhance-best-fit-node-logic
